File: favourite-lyrics-jpa/src/main/java/org/jesperancinha/lyrics/jpa/adapter/LyricsJpaAdapter.java
package org.jesperancinha.lyrics.jpa.adapter;

import lombok.SneakyThrows;
import org.jesperancinha.lyrics.domain.data.LyricsDto;
import org.jesperancinha.lyrics.domain.exception.LyricsNotFoundException;
import org.jesperancinha.lyrics.domain.port.LyricsPersistencePort;
import org.jesperancinha.lyrics.jpa.model.LyricsEntity;
import org.jesperancinha.lyrics.jpa.repository.LyricsRepository;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Objects;
import java.util.UUID;
import java.util.function.Supplier;
import java.util.stream.Collectors;

@Service
public class LyricsJpaAdapter implements LyricsPersistencePort {

    private final LyricsRepository lyricsRepository;

    public LyricsJpaAdapter(LyricsRepository lyricsRepository) {
        this.lyricsRepository = lyricsRepository;
    }

    @Override
    public void addLyrics(LyricsDto lyricsDto) {
        final LyricsEntity lyricsEntity = getLyricsEntity(lyricsDto);
        lyricsRepository.save(lyricsEntity);
    }

    @Override
    public void removeLyrics(LyricsDto lyricsDto) {
        lyricsRepository.deleteAllByParticipatingArtist(lyricsDto.getParticipatingArtist());
    }

    @Override
    public void updateLyrics(LyricsDto lyricsDto) {
        final LyricsEntity byParticipatingArtist = lyricsRepository.findByParticipatingArtist(lyricsDto.getParticipatingArtist());
        if (Objects.nonNull(byParticipatingArtist)) {
            byParticipatingArtist.setLyrics(lyricsDto.getLyrics());
            lyricsRepository.save(byParticipatingArtist);
        } else {
            final LyricsEntity byLyrics = lyricsRepository.findByLyrics(lyricsDto.getLyrics());
            if (Objects.nonNull(byLyrics)) {
                byLyrics.setParticipatingArtist(lyricsDto.getParticipatingArtist());
                lyricsRepository.save(byLyrics);
            }
        }
    }

    @Override
    public List<LyricsDto> getAllLyrics() {
        return lyricsRepository.findAll()
                .stream()
                .map(this::getLyrics)
                .collect(Collectors.toList());
    }

    @SneakyThrows
    @Override
    public LyricsDto getLyricsById(UUID lyricsId) {
        return getLyrics(lyricsRepository.findById(lyricsId)
                .orElseThrow((Supplier<Throwable>) () -> new LyricsNotFoundException(lyricsId)));
    }

    private LyricsEntity getLyricsEntity(LyricsDto lyricsDto) {
        return LyricsEntity.builder()
                .participatingArtist(lyricsDto.getParticipatingArtist())
                .lyrics(lyricsDto.getLyrics())
                .build();
    }

    private LyricsDto getLyrics(LyricsEntity lyricsEntity) {
        return LyricsDto.builder()
                .participatingArtist(lyricsEntity.getParticipatingArtist())
                .lyrics(lyricsEntity.getLyrics())
                .build();
    }

}

File: favourite-lyrics-jpa/src/main/kotlin/org/jesperancinha/lyrics/jpa/adapter/LyricsJpaAdapter.kt
package org.jesperancinha.lyrics.jpa.adapter

import org.jesperancinha.lyrics.domain.data.LyricsDto
import org.jesperancinha.lyrics.domain.data.LyricsFullDto
import org.jesperancinha.lyrics.domain.exception.LyricsNotFoundException
import org.jesperancinha.lyrics.domain.port.LyricsPersistencePort
import org.jesperancinha.lyrics.jpa.model.LyricsEntity
import org.jesperancinha.lyrics.jpa.repository.LyricsRepository
import org.springframework.stereotype.Service
import java.util.*

@Service
class LyricsJpaAdapter(private val lyricsRepository: LyricsRepository) : LyricsPersistencePort {
    override fun addLyrics(lyricsDto: LyricsDto) {
        val lyricsEntity = getLyricsEntity(lyricsDto)
        lyricsRepository.save(lyricsEntity)
    }

    override fun removeLyrics(lyricsDto: LyricsDto) {
        lyricsRepository.deleteAllByParticipatingArtist(requireNotNull(lyricsDto.participatingArtist))
    }

    override fun updateLyrics(lyricsDto: LyricsDto) {
        val byParticipatingArtist =
            lyricsRepository.findByParticipatingArtist(requireNotNull(lyricsDto.participatingArtist))
        if (isOnlyNullValues(byParticipatingArtist)) {
            lyricsRepository.save(byParticipatingArtist.copy(lyrics = lyricsDto.lyrics))
        } else {
            val findByLyrics: LyricsEntity = lyricsRepository.findByLyrics(requireNotNull(lyricsDto.lyrics))
            if (!isOnlyNullValues(findByLyrics)) {
                lyricsRepository.save(findByLyrics.copy(participatingArtist = lyricsDto.participatingArtist))
            }
        }
    }

    override fun getAllLyrics(): List<LyricsDto> = lyricsRepository.findAll()
        .filterNotNull()
        .map { lyricsEntity -> getLyrics(lyricsEntity) }

    override fun getAllLFullLyrics(): List<LyricsFullDto> = lyricsRepository.findAll()
        .filterNotNull()
        .map { lyricsEntity -> fullLyrics(lyricsEntity) }

    override fun getLyricsById(lyricsId: UUID): LyricsDto = getLyrics(
        requireNotNull(lyricsRepository.findById(lyricsId)
            .orElseThrow { LyricsNotFoundException(lyricsId) })
    )

    private fun getLyricsEntity(lyricsDto: LyricsDto): LyricsEntity {
        return LyricsEntity(
            participatingArtist = lyricsDto.participatingArtist,
            lyrics = lyricsDto.lyrics
        )
    }

    private fun getLyrics(lyricsEntity: LyricsEntity): LyricsDto {
        return LyricsDto(
            participatingArtist = lyricsEntity.participatingArtist,
            lyrics = lyricsEntity.lyrics
        )

    }

    private fun fullLyrics(lyricsEntity: LyricsEntity): LyricsFullDto {
        return LyricsFullDto(
            participatingArtist = requireNotNull(lyricsEntity.participatingArtist),
            lyrics = requireNotNull(lyricsEntity.lyrics),
            uuid = requireNotNull(lyricsEntity.id)
        )
    }
}
fun isOnlyNullValues(lyricsEntity: LyricsEntity)
        = lyricsEntity.participatingArtist == null && lyricsEntity.lyrics == null && lyricsEntity.id == null
